tutoraspects/templates/aspects/apps/aspects/migrations/alembic/versions/0009_prefer_course_key.py

```diff
@@ -6,208 +8,430 @@
+        ) as Int16)
+    ) AS starting_position,
+    JSON_VALUE(
+        event_str,
+        '$.context.extensions."http://id.tincanapi.com/extension/ending-point"'
+    ) AS ending_point
+FROM
+    {{ ASPECTS_XAPI_DATABASE }}.{{ ASPECTS_XAPI_TABLE }}
+WHERE verb_id IN (
+    'https://w3id.org/xapi/dod-isd/verbs/navigated'
+);
+"""
+
+NEW_NAVIGATION_QUERY = """
+SELECT
+    event_id,
+    emission_time,
+    actor_id,
+    object_id,
+    splitByString('/', course_id)[-1] AS course_key,
+    org,
+    verb_id,
+    JSON_VALUE(event_str, '$.object.definition.type') AS object_type,
+    -- clicking a link and selecting a module outline have no starting-position field
+    if (
+        object_type in (
+            'http://adlnet.gov/expapi/activities/link',
+            'http://adlnet.gov/expapi/activities/module'
+        ),
+        0,
+        cast(JSON_VALUE(
+            event_str,
+            '$.context.extensions."http://id.tincanapi.com/extension/starting-position"'
+        ) as Int16)
+    ) AS starting_position,
+    JSON_VALUE(
+        event_str,
+        '$.context.extensions."http://id.tincanapi.com/extension/ending-point"'
+    ) AS ending_point
+FROM
+    {{ ASPECTS_XAPI_DATABASE }}.{{ ASPECTS_XAPI_TABLE }}
+WHERE verb_id IN (
+    'https://w3id.org/xapi/dod-isd/verbs/navigated'
+);
+"""
+
+MIGRATIONS = [
+    MvMigration(
+        "{{ ASPECTS_XAPI_DATABASE }}.{{ ASPECTS_ENROLLMENT_EVENTS_TABLE }}",
+        OLD_ENROLLMENT_DDL,
+        NEW_ENROLLMENT_DDL,
+        "{{ ASPECTS_XAPI_DATABASE }}.{{ ASPECTS_ENROLLMENT_TRANSFORM_MV }}",
+        OLD_ENROLLMENT_QUERY,
+        NEW_ENROLLMENT_QUERY,
+    ),
+    MvMigration(
+        "{{ ASPECTS_XAPI_DATABASE }}.{{ ASPECTS_VIDEO_PLAYBACK_EVENTS_TABLE }}",
+        OLD_VIDEO_DDL,
+        NEW_VIDEO_DDL,
+        "{{ ASPECTS_XAPI_DATABASE }}.{{ ASPECTS_VIDEO_PLAYBACK_TRANSFORM_MV }}",
+        OLD_VIDEO_QUERY,
+        NEW_VIDEO_QUERY,
+    ),
+    MvMigration(
+        "{{ ASPECTS_XAPI_DATABASE }}.{{ ASPECTS_PROBLEM_EVENTS_TABLE }}",
+        OLD_PROBLEM_DDL,
+        NEW_PROBLEM_DDL,
+        "{{ ASPECTS_XAPI_DATABASE }}.{{ ASPECTS_PROBLEM_TRANSFORM_MV }}",
+        OLD_PROBLEM_QUERY,
+        NEW_PROBLEM_QUERY,
+    ),
+    MvMigration(
+        "{{ ASPECTS_XAPI_DATABASE }}.{{ ASPECTS_NAVIGATION_EVENTS_TABLE }}",
+        OLD_NAVIGATION_DDL,
+        NEW_NAVIGATION_DDL,
+        "{{ ASPECTS_XAPI_DATABASE }}.{{ ASPECTS_NAVIGATION_TRANSFORM_MV }}",
+        OLD_NAVIGATION_QUERY,
+        NEW_NAVIGATION_QUERY,
+    ),
+]
+
+
+def migrate(table_name, mv_name, table_ddl, mv_query):
+    # for each table and materialized view to migrate:
+    # - drop the existing table and materialized view if they exist
+    # - create the new table
+    # - load data into the new table using the new MV query
+    # - create the materialized view using the new query
+    op.execute(f"DROP TABLE IF EXISTS {table_name}")
+    op.execute(f"DROP TABLE IF EXISTS {mv_name}")
```

Comment:
Does this actually do the drop of the mv? I would expect this to need a `drop view` but maybe fail silently if it doesn't find a table?